PIMbench/filter-by-key/PIM/filter.cpp

```diff
@@ -86,160 +87,152 @@ struct Params getInputParams(int argc, char **argv)
-  {
-    outVector.insert(outVector.end(), outSubVector[i].begin(), outSubVector[i].end());
-  }
-
-  auto stop_cpu = std::chrono::high_resolution_clock::now();
-  hostElapsedTime += (stop_cpu - start_cpu);
 
-  // run scan on host
-  filterByKey(inVector, inVectorSize, p.key, bitMapHost);
-
-  // select data whose bitmap is '1' on host based on bitmap from host
-  for (uint64_t i = 0; i < inVectorSize; i++)
-  {
-    if (bitMapHost[i] == 1)
-    {
-      outVectorHost.push_back(inVector[i]);
-    }
-  }
-  // !check results and print it like km
-  if (p.shouldVerify)
-  {
-    bool ok = true;
-    if (outVector != outVectorHost)
-    {
-      std::cout << "Wrong answer!" << std::endl;
-      ok = false;
-    }
-    if (ok)
-    {
-      std::cout << "All correct!" << std::endl;
+    // // uint64_t outSize = selectedNum;
+    uint64_t buffer_in_CPU = 0;
+    uint64_t buffer_in_CPU1 = 0;
+    uint64_t buffer_in_CPU2 = 0;
+    uint64_t buffer_in_CPU3 = 0;
+    uint64_t buffer_in_CPU4 = 0;
+    uint64_t buffer_in_CPU5 = 0;
+    uint64_t buffer_in_CPU6 = 0;
+    uint64_t buffer_in_CPU7 = 0;
+    uint64_t buffer_in_CPU8 = 0;
+    uint64_t selectedNum1 = 0;
+    uint64_t selectedNum2 = 0;
+    uint64_t selectedNum3 = 0;
+    uint64_t selectedNum4 = 0;
+    uint64_t selectedNum5 = 0;
+    uint64_t selectedNum6 = 0;
+    uint64_t selectedNum7 = 0;
+    uint64_t selectedNum8 = 0;
+
+    start_cpu = std::chrono::high_resolution_clock::now();
+    for (uint64_t i = 0; i < inVectorSize; i+=8){
+      _mm_prefetch((char *)&bitMapTemp[i + 512], _MM_HINT_T0);
+      if(bitMapTemp[i + 0] == true){
+        buffer_in_CPU1 += inVector[i + 0];
+        selectedNum1++;
+      }
+      if(bitMapTemp[i + 1] == true){
+        buffer_in_CPU2 += inVector[i + 1];
+        selectedNum2++;
+      }
+      if(bitMapTemp[i + 2] == true){
+        buffer_in_CPU3 += inVector[i + 2];
+        selectedNum3++;
+      }
+      if(bitMapTemp[i + 3] == true){
+        buffer_in_CPU4 += inVector[i + 3];
+        selectedNum4++;
+      }
+      if(bitMapTemp[i + 4] == true){
+        buffer_in_CPU5 += inVector[i + 4];
+        selectedNum5++;
+      }
+      if(bitMapTemp[i + 5] == true){
+        buffer_in_CPU6 += inVector[i + 5];
+        selectedNum6++;
+      }
+      if(bitMapTemp[i + 6] == true){
+        buffer_in_CPU7 += inVector[i + 6];
+        selectedNum7++;
+      }
+      if(bitMapTemp[i + 7] == true){
+        buffer_in_CPU8 += inVector[i + 7];
+        selectedNum8++;
+      }
     }
-  }
-  pimShowStats();
+    uint64_t outSize = selectedNum1 + selectedNum2 + selectedNum3 + selectedNum4 + selectedNum5 + selectedNum6 + selectedNum7 + selectedNum8;
 
-  cout << "Host elapsed time: " << std::fixed << std::setprecision(3) << hostElapsedTime.count() << " ms." << endl;
-  return 0;
-}
+    auto stop_cpu = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double, std::milli> hostElapsedTime = (stop_cpu - start_cpu);
```

Comment:
Is it possible to move this code block to a function call?

Reply:
It is a performance-specific optimization, I think it is safer to just keep it locally in the main function, and I think the code also runs faster in this way.